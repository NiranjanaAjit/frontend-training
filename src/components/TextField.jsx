import { forwardRef, useEffect, useRef, useState } from "react";
import React from "react";
const TextField = forwardRef((props, ref) => {
  // useEffect(()=>{
  //   props.onChange(id,props.value)
  // },[])
  const onChange = (e) => {
    if (props.onChange) props.onChange(e.target.name, e.target.value);
  };

  let hidden = false;
  let defaultValue = props.value;
  let disabledValue = false;
  if (props.default) {
    disabledValue = true;
    defaultValue = props.default;
  } else if (props.default == "") {
    hidden = true;
  }

  return (
    <span className={props.divClassName} hidden={hidden}>
      <label htmlFor="uname" data-testid="textfield-label-test-id">{props.label}</label>
      <input
        data-testid="textfield-input-test-id"
        type={props.type}
        ref={ref}
        name={props.name}
        id="uname"
        value={props.value}
        placeholder={props.label}
        disabled={disabledValue}
        onChange={onChange}
      />
    </span>
  );
});

export default TextField;
