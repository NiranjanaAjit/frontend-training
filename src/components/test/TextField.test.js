import React from "react";
import TextField from "../TextField";
import { fireEvent, getByTestId, getByText, render, within } from "@testing-library/react";
describe("Text Field Test", () => {
  test("Rendering text field", () => {
    const { getByTestId } = render(<TextField />);
    const element = getByTestId("textfield-input-test-id");
    expect(element).toBeTruthy();
  });

  test("Check if label is correct", () => {
    const label = "test-label";
    const { getByTestId } = render(<TextField label={label} />);
    const element = getByTestId('textfield-label-test-id')
    const {getByText} = within(element);
    expect(getByText(label)).toBeTruthy();
  });

  test("Check if displayed value is correct", () => {
    const value = "test-value";
    const { getByDisplayValue } = render(<TextField value={value} />);
    getByDisplayValue("test-value");
  });

  test("Check if placeholder is correct", () => {
    const placeholder = "input placeholder";
    const { getByPlaceholderText } = render(<TextField label={placeholder} />);
    getByPlaceholderText("input placeholder");
  });

  test("Check if value is changed on typing", () => {
    const { getByTestId } = render(<TextField />);
    const element = getByTestId("textfield-input-test-id");
    fireEvent.change(element, { target: { value: "test-change" } });
    expect(element.value).toBe("test-change");
  });

  test("Check if onChange is triggered", () => {
    const onChange = jest.fn();
    const { getByTestId } = render(<TextField onChange={onChange} />);
    const element = getByTestId("textfield-input-test-id");
    console.log(element.value); //before change
    fireEvent.change(element, { target: { value: "test-onChange" } });
    console.log(element.value); //after change
    expect(onChange).toHaveBeenCalledTimes(1);
  });

  test("Check if snapshots are mapped properly", () => {
    const onChange = jest.fn();
    const text = "bye";
    const { asFragment } = render(
      <TextField label={text} value="test-snapshot" />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
