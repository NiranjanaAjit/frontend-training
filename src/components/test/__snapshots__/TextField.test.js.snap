// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Field Test Check if snapshots are mapped properly 1`] = `
<DocumentFragment>
  <span>
    <label
      data-testid="textfield-label-test-id"
      for="uname"
    >
      bye
    </label>
    <input
      data-testid="textfield-input-test-id"
      id="uname"
      placeholder="bye"
      value="test-snapshot"
    />
  </span>
</DocumentFragment>
`;
